gu-client/src/async.rs

```diff
@@ -0,0 +1,297 @@
+#![allow(dead_code)]
+
+use actix_web::{client, HttpMessage};
+use bytes::Bytes;
+use error::Error;
+use futures::stream::Stream;
+use futures::{future, Future};
+use gu_net::rpc::peer::PeerInfo;
+use std::str;
+use std::sync::Arc;
+
+/// Hub session information.
+#[derive(Clone, Serialize, Deserialize, Debug, Default, Builder)]
+#[builder(pattern = "owned", setter(into))]
+pub struct SessionInfo {
+    name: String,
+    environment: String,
+}
+
+/// Peer session information.
+#[derive(Clone, Serialize, Deserialize, Debug, Default, Builder)]
+#[builder(pattern = "owned", setter(into))]
+pub struct PeerSessionInfo {
+    image: PeerSessionImage,
+    environment: String,
+}
+
+/// Peer session image.
+#[derive(Clone, Serialize, Deserialize, Debug, Default)]
+pub struct PeerSessionImage {
+    url: String,
+    hash: String,
+}
+
+/// Connection to a single hub.
+#[derive(Clone, Debug)]
+pub struct Driver {
+    driver_inner: Arc<DriverInner>,
+}
+
+#[derive(Debug)]
+struct DriverInner {
+    url: String,
+}
+
+impl Driver {
+    /// creates a driver from a given address:port, e.g. 127.0.0.1:61621
+    pub fn from_addr<T: Into<String>>(addr: T) -> Driver {
```

Comment:
we may consider using `url::Url` to store the url to early detect if the supplied url is invalid.

Reply:
Done.